kernel-args: Update for cgroups v2 by default (#269)
coreos/fedora-coreos-docs: modules/ROOT/pages/kernel-args.adoc

= Modifying Kernel Arguments

Kernel arguments changes are managed by `rpm-ostree` via the https://www.mankier.com/1/rpm-ostree[`rpm-ostree kargs`] sub command. Changes are applied to a new deployment and a reboot is necessary for those to take effect.

.Append kernel argument; useful with e.g. console= that can be used multiple times. empty value for an argument is allowed
[source,bash]
----
$ sudo rpm-ostree kargs --append=KEY=VALUE
----

.Replace existing kernel argument, the user is also able to replace an argument with KEY=VALUE if only one value exist for that argument
[source,bash]
----
$ sudo rpm-ostree kargs --replace=KEY=VALUE=NEWVALUE
----

.Delete a specific kernel argument key/val pair or an entire argument with a single key/value pair
[source,bash]
----
$ sudo rpm-ostree kargs --delete=KEY=VALUE
----

.Use an editor to modify the kernel arguments
[source,bash]
----
$ sudo rpm-ostree kargs --editor
----

.Exemple: Update an existing system from cgroupsv1 to cgroupsv2 and immediately reboot

[source,bash]
----
$ sudo rpm-ostree --delete=systemd.unified_cgroup_hierarchy --reboot

> ----

Reviewer: This seems like a bad example now because we're defaulting to cgroups v2 so `systemd.unified_cgroup_hierarchy=0` won't be on the kernel command line. Is there another example we can use?

Author: We can add another example but we still need a place to put this one for all our existing users that will have to manually update their systems.

Author: Added more examples!

Reviewer: > We can add another example but we still need a place to put this one for all our existing users that will have to manually update their systems.

Good point. I didn't think of that.

= Modifying Kernel Arguments

> Kernel arguments changes are managed by `rpm-ostree` via the https://www.mankier.com/1/rpm-ostree[`rpm-ostree kargs`] sub command. Changes are applied to a new deployment and a reboot is necessary for those to take effect.

Reviewer 2: Aside: feels funny to link to a third-party rendering of our own man pages. Maybe we should figure out a way to render them in https://coreos.github.io/rpm-ostree/.

Author: Yep!

== Adding kernel arguments

> You can append kernel argument. This is useful with e.g. `console=` that can be used multiple times. An empty value for an argument is allowed:

Reviewer 2: ```suggestion
You can append kernel arguments. This is useful with e.g. `console=` that can be used multiple times. An empty value for an argument is allowed:
```

[source,bash]
----
$ sudo rpm-ostree kargs --append=KEY=VALUE
----

> .Exemple: Add reserved memory for Kdump support

Reviewer 2: ```suggestion
.Example: Add reserved memory for Kdump support
```

(and below :) )

[source,bash]
----
$ sudo rpm-ostree kargs --append='crashkernel=256M'
----

See also xref:debugging-kernel-crashes.adoc[Debugging kernel crashes using kdump].

== Removing existing kernel arguments

You can delete a specific kernel argument key/value pair or an entire argument with a single key/value pair:

[source,bash]
----
$ sudo rpm-ostree kargs --delete=KEY=VALUE
----

.Exemple: Remove console parameters to enable kernel auto-detection

[source,bash]
----
$ sudo rpm-ostree kargs --delete 'console=ttyS0,115200n8'
----

See also xref:emergency-shell.adoc[Emergency console access].

.Exemple: Update an existing system from cgroupsv1 to cgroupsv2 and immediately reboot

[source,bash]
----
$ sudo rpm-ostree kargs --delete=systemd.unified_cgroup_hierarchy --reboot
----

== Replacing existing kernel arguments

> You can replace an existing kernel argument with a new value. You can direclty use `KEY=VALUE` if only one value exist for that argument:

Reviewer 2: ```suggestion
You can replace an existing kernel argument with a new value. You can directly use `KEY=VALUE` if only one value exists for that argument. Otherwise, you can specify the new value using the following format:
```

== Interactive editing

To use an editor to modify the kernel arguments:

[source,bash]
----
$ sudo rpm-ostree kargs --editor
----

== Modifying Kernel Arguments via Ignition

Currently to change kernel arguments via Ignition, you must script a systemd service which runs `rpm-ostree kargs` and then trigger a reboot.

NOTE: In the future, we will have a more Ignition-friendly method of doing this with stronger guarantees. See upstream issues https://github.com/coreos/ignition/issues/1051[ignition#1051], https://github.com/coreos/fedora-coreos-tracker/issues/318[fedora-coreos-tracker#318] and https://github.com/coreos/fedora-coreos-tracker/issues/752[fedora-coreos-tracker#752] for more information.

NOTE:  The `After=systemd-machine-id-commit.service` directive is important in the following examples to avoid some subtle issues. Similarly, any `ConditionFirstBoot=true` services should use `Before=first-boot-complete.target systemd-machine-id-commit.service`. See https://github.com/systemd/systemd/blob/3045c416e1cbbd8ab40577790522217fd1b9cb3b/man/systemd.unit.xml#L1315[the systemd documentation] for more details.

=== Example: Staying on cgroups v1

> Cgroups v2 is now the default on Fedora CoreOS. Here's an example which adds the `systemd.unified_cgroup_hierarchy=0` kernel argument so that the machine keeps using cgroups v1:

Reviewer 2: ```suggestion
Cgroups v2 is now the default on Fedora CoreOS on the `next` stream. Here's an example which adds the `systemd.unified_cgroup_hierarchy=0` kernel argument so that the machine keeps using cgroups v1:
```

        - $pubkey
systemd:
  units:
    - name: cgroups-v1-karg.service
      enabled: true
      contents: |
        [Unit]

>         Description=Switch back to cgroups v1

Reviewer: ```suggestion
        Description=Switch to cgroups v1
```

        # We run after `systemd-machine-id-commit.service` to ensure that
        # `ConditionFirstBoot=true` services won't rerun on the next boot.
        After=systemd-machine-id-commit.service
        ConditionKernelCommandLine=!systemd.unified_cgroup_hierarchy
        ConditionPathExists=!/var/lib/cgroups-v1-karg.stamp

        [Service]
        Type=oneshot
        RemainAfterExit=yes
        ExecStart=/bin/rpm-ostree kargs --append=systemd.unified_cgroup_hierarchy=0
        ExecStart=/bin/touch /var/lib/cgroups-v1-karg.stamp
        ExecStart=/bin/systemctl --no-block reboot

        [Install]
        WantedBy=multi-user.target
----

=== Example: Disabling all CPU vulnerability mitigations

> Here's an example which disables all CPU vulnerability mitigations:

Reviewer: ```suggestion
Here's an example which switches `mitigations=auto,nosmt` to `mitigations=off` to disable all CPU vulnerability mitigations:
```

The reason for spelling it out more is because the formatting for `rpm-ostree kargs replace` is a bit awkward and not 100% intuitive what is getting replaced with what.

        WantedBy=multi-user.target
----

=== Example: Staying on cgroups v1

> Cgroups v2 is now the default on Fedora CoreOS on the `next` stream. Here's an example which adds the `systemd.unified_cgroup_hierarchy=0` kernel argument so that the machine keeps using cgroups v1:

Reviewer 2: Sorry, suggesting one last tweak for this:

```suggestion
Starting from April 13 2021, Cgroups v2 is the default on Fedora CoreOS on the `next` stream. Here's an example which adds the `systemd.unified_cgroup_hierarchy=0` kernel argument so that the machine keeps using cgroups v1:
```

That way, the docs don't lie until next week. :)
